### PR TITLE
Implementation of Read-Only, Write-Only and Required Properties

### LINKED ISSUE #226: Implementing readOnly, writeOnly and Required Properties
### I'm submitting a 
- [ ] bug report.
- [X] feature request.

### Current Behaviour:

Right now although the doc contains properties like readOnly ,writeOnly , Required the functionality is not implemented in Hydrus. We need to implement this functionality so that we can have checksums instead of pk's or UUID's , advantage of that would be that client can reconstruct the checksum to check what is to be fetched .
### Do you want to work on this issue?

~YES~

EDITED BY Mec-iS

### PR DESCRIPTION
**Pull request by aribalam**
Fixes #226 

### Checklist
- [x] My branch is up-to-date with upstream/develop branch.
- [x] Everything works and tested for Python 3.5.2 and above.

### Change logs
1. Changed the database schema to store read-only, write-only and required attributes of a property.
2. Added code for checking the validity of properties during crud operations.
3. Added Exceptions to handle errors related to it.

### Description
Feature to handle read-only, write-only and required attributes of the properties.

The following approach has been made:
1. The database schema has been changed. The read-only, write-only and required attributes of a property are stored in the database.
2. `crud.py` is responsible for checking the validity of properties during crud operations and to handle them accordingly.
3. During get operation, the properties that are write-only are omitted.
4. During insert operation, the object provided are checked for all required properties and exception is raised when any of it is missing.
5. During multiple insertion, only those objects are inserted which has all the required properties, the others are omitted and exception is raised.
6. During update, the object invokes the insert function with the parameter `to_update` set to `True`. This signals the latter to check for read-only properties and raise exception accordingly.
7. Three exceptions has been added. They are as follows - 
`RequiredPropertyNotFound` - When a required property of an object is missing during insert operations.
`InsertMultipleObjectError` - When any object has a missing required property, during insertion of multiple objects.
`ReadOnlyException` - When the object given by the user to update an existing object has a read-only property.

The following problem still persists -
1. `crud.py` uses `get_doc()` to access nested class paths, achieve segregation of objects and check for the required properties of objects. This is generally not a good design as `crud.py` should only be accessing the database. We might need to change the database schema in the future to store classes and their properties.
2. crud tests could not run as sample docs could not be passed on to `crud.py`

### CODE REVIEW

**Comment by aribalam** [comment]
> how does it compare to #326 ?

@Mec-iS. Comparing with #326, there are certain differences.
1. The database schema has been changed to include read-only, write-only and required properties.
2. All checks for the 3 attributes are only done by `crud.py`.
3. No HydraDoc objects are passed on to crud functions.
4. `crud_tests` are still not able to run because of the above point.

> could you work with the other developers to provide an implementation that puts together best of both PRs?

I will be happy to. @vddesai1871  We can have a discussion to make a best out of both?

**Comment by vddesai1871** [comment]
@aribalam Have you ran other tests locally?
I am having trouble manually testing read-only properties...

**Comment by aribalam** [comment]
The read-only property was not tested directly as there is a bug in update function that doesn't delete the instance properly from the database. I will be posting a new issue for that. But instead, I tested `crud.insert` with `to_update` set to True which actually does the checking. It was working just fine.

File: hydrus/data/crud.py [hunk 1]
```diff
@@ -182,6 +193,27 @@ def insert(object_: Dict[str, Any], session: scoped_session,
     session.add(instance)
     session.flush()
 
+    doc = get_doc()
```
[the remark this answers is not in the document]
Comment (chrizandr) [reply hydrus/data/crud.py]: Also using get_doc without a running server will not work, the flask object relies on session data to store the API Doc. We want to keep CRUD as decoupled from the server as possible, kind of like a data manipulation API that can be used independently.

Maybe both of you can discuss on a workaround that incorporates both your work?

**Comment by aribalam** [comment]
How about we use the following approach.
1. We create a db model that maps the classes with its properties. That way, we will surely be able to check for required properties.
2. For nested class paths, we add a column in `RDFClass` that states whether a class is a non-collection or a collection class. We can achieve this by using polymorphic identities on `RDFClass`, similar to what has been done with `BaseProperty` model. Minor changes will to be made in `doc_parse` functions and `main.py`. 

This will greatly reduce the code in `crud.py` and can even ease implementation for future objectives.It will also make `crud.py` independent of ApiDoc.

@vddesai1871 @chrizandr @xadahiya any suggestions ?

**Comment by chrizandr** [comment]
> How about we use the following approach.
> 
> 1. We create a db model that maps the classes with its properties. That way, we will surely be able to check for required properties.

The thing is that these things are already available in the API Documentation, defining them again in the DB would be both redundant and constraining. Defining relations from the API documentation would mean that the data in the database would be used only by that API Doc, we can't use the same data for any other API Documentation. The main use case of Hydra is that the server should be usable without the need to hard code anything, so if the API changes the client would be able to adapt to it. I think this use case would be compromised if we need to redefine our data and database every time we change the structure of the server/API Documentation.
We should also consider the case where we use OpenAPI specs to create the server. In such cases, the API might not be replicated as intended, since there is no perfect mapping between a Hydra server and the OpenAPI server. Here there would be a need to change the server structure repeatedly as and when new use cases are added, redefining the database every time would be an expensive operation, specially for large databases.

This is my two cents anyway. Why don't we open a PR about this and discuss further, I will add work in progress tags to this meanwhile.

**Comment by aribalam** [comment]
@chrizandr Thanks for clearing this out. :)
I think we should have another layer in between `views` (or `resources`) and `crud.py`. This layer would be able to access the APIDoc. It should be responsible for performing all sorts of validations and processing on requests (from `resource.py`) and responses (from `crud.py`).

The workflow would be as follows - 

`resource ---(object)---> validation layer(using doc) ---(processed object)---> crud`
`crud ---(response)---> validation layer(using doc) ---(processed response)---> resource`

**Comment by xadahiya** [comment]
I think we should move forward with #326. Consider the scenario where the HydraDoc is changed slightly only changing several property types. Is it a good idea to change the database each time. I agree with @vddesai1871 on this. His approach seems much clearer.

@chrizandr ^^

**Comment by chrizandr** [comment]
Yes, that may seem to be the case, we should move forward with changing HydraDoc to do all the checks. It would make code much cleaner.
I'll close this PR for now.

**Pull request closed without merging**